### DISCUSSION TITLE
unknown "mage_run_type" variable - need help with multi-store setup

### DISCUSSION THREAD

**Discussion by giorgi17** [markshust/docker-magento#1036] [Q&A]
Following “Set Up a Magento 2 Development Environment with Docker” course.
![image](https://github.com/markshust/docker-magento/assets/21131956/747eb2e2-6262-4f84-bf7b-221d385f988d)

This is my `/images/nginx/conf/default.conf`
![image](https://github.com/markshust/docker-magento/assets/21131956/cc66d4c2-d32a-41ed-96a9-70f726d9d1ee)
`upstream fastcgi_backend {
  server unix:/sock/docker.sock;
}

map $http_host $MAGE_RUN_CODE {
    default base;
    foo.test foo_website;
}

server {
  listen 8000;
  return 301 https://$host$request_uri;
}

server {
  listen [::]:8443 ssl http2 ipv6only=on;
  listen 8443 ssl http2;

  ssl_certificate /etc/nginx/certs/nginx.crt;
  ssl_certificate_key /etc/nginx/certs/nginx.key;

  server_name magento.test foo.test;

  set $MAGE_ROOT /var/www/html;
  set $MAGE_RUN_TYPE website;

  fastcgi_buffer_size 64k;
  fastcgi_buffers 8 128k;

  location /livereload.js {
    proxy_set_header Host $host;
    proxy_pass http://phpfpm:35729/livereload.js;
  }

  location /livereload {
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_pass http://phpfpm:35729/livereload;
  }

  include /var/www/html/nginx[.]conf;
}`

nginx.conf file in Magento project:
![image](https://github.com/markshust/docker-magento/assets/21131956/8b85bb83-beea-431e-8aad-d697c0f2ed01)

**Comment by JesperHerrloff** [comment]
How are your nginx.conf look in magento project folder. Is it lowercased $mage_run_type?

**Reply by giorgi17** [reply]
I've updated the question and you can also see that no, it's not lowercased.

**Reply by rubenromao** [reply]
hi @giorgi17 did you managed to set it up?
I'm facing issues adding sub-domains!

**Reply by giorgi17** [reply]
> hi @giorgi17 did you managed to set it up?
> 
> I'm facing issues adding sub-domains!

No, not really, gave up a while ago.

**Reply by rubenromao** [reply]
I managed to get it working.

**Reply by giorgi17** [reply]
What was the issue?

**Reply by rubenromao** [reply]
Here's my configuration for 3 websites:

1 - Stop the containers.


2 - Change "compose.yaml" nginx:
```
services:
  app:
#    image: markoshust/magento-nginx:1.18-8
    build: images/nginx
```


3 - Change "compose.dev.yaml" to include "nginx.conf":
```
#      - ./src/nginx.conf.sample:/var/www/html/nginx.conf:cached
      - ./src/nginx.conf:/var/www/html/nginx.conf:delegated
```


4 - Create "Doxkerfile" under "images/nginx/Dockerfile" with this content:
```
FROM markoshust/magento-nginx:1.18-8

COPY --chown=app:app ./conf/default.conf /etc/nginx/conf.d/
```


5 - Create "default.conf" under "images/nginx/conf/default.conf" and "Doxkerfile" :
```
upstream fastcgi_backend {
    server unix:/sock/docker.sock;
}

map $http_host $MAGE_RUN_CODE {
    default base;
    domain.local base;
    subdomain1.domain.local mageStoreCodeHere;
    subdomain2.domain.local mageStoreCodeHere;
}

server {
    listen 8000;
    return 301 https://$host$request_uri;
}

server {
    listen [::]:8443 ssl http2 ipv6only=on;
    listen 8443 ssl http2;

    ssl_certificate /etc/nginx/certs/nginx.crt;
    ssl_certificate_key /etc/nginx/certs/nginx.key;

    server_name domain.local subdomain1.domain.local subdomain2.domain.local;

    set $MAGE_ROOT /var/www/html;
    set $MAGE_RUN_TYPE website;

    fastcgi_buffer_size 64k;
    fastcgi_buffers 8 128k;

    include /var/www/html/nginx[.]conf;
}
```


6 - Change "nginx.conf" PHP entry like this:
```
# PHP entry point for main application
location ~ ^/(index|get|static|errors/report|errors/404|errors/503|health_check)\.php$ {
    try_files $uri =404;
    fastcgi_pass   fastcgi_backend;

    fastcgi_param  PHP_FLAG  "session.auto_start=off \n suhosin.session.cryptua=off";
    fastcgi_param  PHP_VALUE "memory_limit=756M \n max_execution_time=18000";
    fastcgi_read_timeout 600s;
    fastcgi_connect_timeout 600s;

    fastcgi_index  index.php;
    fastcgi_param  SCRIPT_FILENAME  $document_root$fastcgi_script_name;
    fastcgi_param MAGE_RUN_TYPE $MAGE_RUN_TYPE;
    fastcgi_param MAGE_RUN_CODE $MAGE_RUN_CODE;

    include        fastcgi_params;
}
```

7 - Start the containers "bin/start"
I strongly recommend to remove the existing nginx images volumes and containers before start.


8 - Setup the domains (I ran for all of them but tbh I don't know if it is needed)
`bin/setup-domain domain.local`
`bin/setup-domain subdomain1.domain.local`
`bin/setup-domain subdomain2.domain.local`


9 - Change your database URLs by replacing the base URLs for each website scope with your domains:
```bin/n98-magerun2 db:query "UPDATE `core_config_data` SET `value` = REPLACE(`value`, 'website1, 'domain.local')"```
```bin/n98-magerun2 db:query "UPDATE `core_config_data` SET `value` = REPLACE(`value`, 'website2, 'subdomain1.domain.local')"```
```bin/n98-magerun2 db:query "UPDATE `core_config_data` SET `value` = REPLACE(`value`, 'website3, 'subdomain2.domain.local')"```

10 - Flush the cache
`bin/magento c:f`

That's it.
I think I haven't miss anything.

Cheers
